src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -23,6 +26,98 @@ public static class Renamer
         }
 
+        /// <summary>
+        /// Similar to calling <see cref="Document.WithName(string)" /> with additional changes to the solution. 
+        /// Each change is added as a <see cref="RenameDocumentAction"/> in the returned <see cref="RenameDocumentActionSet.ApplicableActions" />.
+        /// 
+        /// Each action may individually encounter errors that prevent it from behaving correctly. Those are reported in <see cref="RenameDocumentAction.GetErrors(System.Globalization.CultureInfo?)"/>.
+        /// 
+        /// Current supported actions that may be returned: 
+        /// * <see cref="RenameSymbolDocumentAction"/> that will rename the type to match the document name
+        /// </summary>
+        public static async Task<RenameDocumentActionSet> RenameDocumentNameAsync(
+            Document document,
+            string newDocumentName,
+            OptionSet optionSet,
+            CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (newDocumentName == null)
+            {
+                throw new ArgumentNullException(nameof(newDocumentName));
+            }
+
+            using var _ = ArrayBuilder<RenameDocumentAction>.GetInstance(out var actions);
+
+            if (!newDocumentName.Equals(document.Name))
+            {
+                var renameAction = await RenameSymbolDocumentAction.TryCreateAsync(document, newDocumentName, cancellationToken).ConfigureAwait(false);
+
+                if (renameAction is object)
+                {
+                    actions.Add(renameAction);
+                }
+            }
+
+            return new RenameDocumentActionSet(
+                actions.ToImmutable(),
+                document.Project.Id,
+                document.Id,
+                newDocumentName,
+                document.Folders,
+                optionSet);
+        }
+
+        /// <summary>
+        /// Similar to calling <see cref="Document.WithFolders(IEnumerable{string})" /> with additional changes to the solution. 
+        /// Each change is added as a <see cref="RenameDocumentAction"/> in the returned <see cref="RenameDocumentActionSet.ApplicableActions" />.
+        /// 
+        /// Each action may individually encounter errors that prevent it from behaving correctly. Those are reported in <see cref="RenameDocumentAction.GetErrors(System.Globalization.CultureInfo?)"/>.
+        /// 
+        /// Current supported actions that may be returned: 
+        /// * <see cref="SyncNamespaceDocumentAction"/> that will sync the namespace(s) of the document to match the document folders
+        /// </summary>
+        public static async Task<RenameDocumentActionSet> RenameDocumentFoldersAsync(
+            Document document,
+            IReadOnlyList<string> newFolders,
+            OptionSet optionSet,
```

Comment:
1. feels like optionSet should be optional on a public API.
2. i'm not sure how i feel about the inconsistency here betwene taking in IReadOnlyLists but returning immutablearrays.  @jasonmalinowski ?

Reply:
I'm not sure we actually have a clear API guidance here. For me the options for accepting things in mostly seems to be:

1. ImmutableArray. Good and immutable, but unless you already have one requires a single array allocation to convert whatever you have into that.
2. IReadOnlyList: doesn't guarantee immutability of the input (it could be modified underneath us) but you can pass in plenty of things.
3. IEnumerable: has to be enumerated just once to be careful, and has least guarantees of sanity. Almost always requires us to reenumerate it into some other type internally.

This is at least consistent with the type returned from Document.Folders so if you were getting a list from that way you can just pass it through.

Returning ImmutableArray: seems appropriate.

Reply:
OptionSet wasn't optional only because the other rename APIs didn't have it as optional. I'm happy to change if that's what we think is correct, it's just there for consistency

Reply:
yeah, it should be optional.